atmost/atmost.man
.TH ATMOST 1 "2019-08-06" "Limits concurrent execution of executable."
.SH "NAME"
atmost \- tool for limiting the concurrent execution of a specified executable(s).
.SH "SYNOPSIS"
.SY atmost
.OP \-v|\-\-verbose
.OP \-f|\-\-file file\-path
.OP \-s|\-\-socket socket\-path
.OP \-r|\-\-reset
.OP \-i|\-\-sleep\-interval seconds
.OP \-c|\-\-print\-current
.OP \-C|\-\-print\-all\-current
.OP \-n|\-\-max\-instances count
.OP \-l|\-\-max\-cpu\-load\-1m percent
.OP \-L|\-\-max\-cpu\-load\-5m percent
.OP \-m|\-\-min\-avail\-ram percent
.OP \-M|\-\-min\-free\-ram percent
.OP \-S|\-\-min\-free\-swap  percent
.OP \-p|\-\-max\-total\-procs count
.OP \-tc|\-\-max\-cpu\-temp temp
.OP \-tg|\-\-max\-gpu\-temp temp
.OP \-tb|\-\-max\-bat\-temp temp
.OP \-io|\-\-max\-io\-ops count
.OP \-nw|\-\-max\-nw\-speed speed
.OP \-snw|\-\-slow\-nw\-speed speed
[\-\- \fIexecutable\fR [args\.\.\.]]
.YS
.
.SY atmost
.B -h
.SY atmost
.B --help
.YS
.SH "DESCRIPTION"
The \fIatmost\fR utility allows to limit concurrent execution of a single
executable or a set of executables, depending on various criteria like
CPU load, memory and swap usage, thermal zone temeratures, current number
of I/O operations, etc., or just by a simple maximum number.

.SH "OPTIONS"
.SS "General options"
.TP
\fB-h\fR, \fB--help\fR
Show this help message and exit.
.TP
\fB-v\fR, \fB--verbose\fR
Make output more verbose. Can be specified multiple times, each instance
increases the verbosity level.
.TP
\fB-f\fR \fIfile-path\fR, \fB--file\fR \fIfile-path\fR
Without this option \fBatmost\fR uses the real path of \fIexecutable\fR
to determine the number of concurrent running instances of the executable.
If the -f option is specified, then \fIfile-path\fR is used instead as a token
used to count concurrent processes. This means that different executables
may be subject of a single counter. \fIfile-path\fR must be a path
to an existing file.
.TP
\fB-s\fR \fIsocket-path\fR, \fB--socket\fR \fIsocket-path\fR
Connect to a local \fBatmost_driver\fR server which will determine when
to start executing the \fIexecutable\fR depending on the logic implemented
by the driver callbacks script.
.TP
\fB-r\fR, \fB--reset\fR 
Reset the instance counter associated with \fIexecutable\fR or with the
\fIfile-path\fR specified after the -f option.
.TP
\fB-i\fR, \fB--sleep-interval\fR
Specifies the time interval in seconds \fBatmost\fR sleeps before rechecking
if the system is overloaded.
.TP
\fB-c\fR, \fB--print-current\fR
Print the current values of the watched execution limits.
.TP
\fB-C\fR, \fB--print-all-current\fR
Print the current values of all supported execution limits.

.SS "Execution limits"
.TP
\fB-n\fR \fIcount\fR, \fB--max-countn\fR \fIcount\fR
Limit the number of instances of \fIexecutable\fR to at most the value
\fIcount\fR, which must be a positive integer.
.TP
\fB-l\fR \fIpercent\fR, \fB-max-cpu-load-1m\fR \fIpercent\fR
Wait before starting a new instance of \fIexecutable\fR while 1 minute average
CPU load is over the value of \fIpercent\fR, which must be a positive
floating-point value,
.TP
\fB-L\fR \fIpercent\fR, \fB-max-cpu-load-5m\fR \fIpercent\fR
Wait before starting a new instance of \fIexecutable\fR while 5 minute average
CPU load is over the value of \fIpercent\fR, which must be a positive
floating-point
value.
.TP
\fB-m\fR \fIpercent\fR, \fB--min-avail-ram\fR \fIpercent\fR
Wait before starting a new instance of \fIexecutable\fR while current amount
of available memory is lower than the value of \fIpercent\fR, which must be
a positive floating-point value.
.TP
\fB-M\fR \fIpercent\fR, \fB--min-free-ram\fR \fIpercent\fR
Wait before starting a new instance of \fIexecutable\fR while current amount
of free memory is lower than the value of \fIpercent\fR, which must be
a positive floating-point value.
.TP
\fB-S\fR \fIpercent\fR, \fB--min-free-swap\fR \fIpercent\fR
Wait before starting a new instance of \fIexecutable\fR while current amount
of free swap space is lower than the value of \fIpercent\fR, which must be
a positive floating-point value.
.TP
\fB-p\fR \fIcount\fR, \fB--max-total-procs\fR \fIcount\fR
Wait before starting a new instance of \fIexecutable\fR while current number
of all processes exceeds the value of \fIcount\fR, which  must be
a positive integer.
.TP
\fB-tc\fR \fItemp\fR, \fB--max-cpu-temp\fR \fItemp\fR
Wait before starting a new instance of \fIexecutable\fR while current CPU
thermal zone temperature exceeds the value of \fItemp\fR.
.TP
\fB-tg\fR \fItemp\fR, \fB--max-gpu-temp\fR \fItemp\fR
Wait before starting a new instance of \fIexecutable\fR while current GPU
thermal zone temperature exceeds the value of\fItemp\fR. 
.TP
\fB-tb\fR \fItemp\fR, \fB--max-bat-temp\fR \fItemp\fR
Wait before starting a new instance of \fIexecutable\fR while current BAT
(battery) thermal zone temperature exceeds the value of\fItemp\fR 
.TP
\fB-io\fR \fIcount\fR, \fB--max-io-ops\fR \fIcount\fR
Wait before starting a new instance of \fIexecutable\fR while current number
of I/O operations exceeds the value of \fIcount\fR. 
.TP
\fB-nw\fR \fIspeed\fR, \fB--min-nw-speed\fR \fIspeed\fR
Wait before starting a new instance of \fIexecutable\fR while current combined
rated speed in MB/s of all network interfaces is less than the value
of \fIspeed\fR.

.SS "Other limits"

.TP
\fB-snw\fR \fIspeed\fR, \fB--slow-nw-speed\fR \fIspeed\fR
Specifies the total rated speed in MB/s of all network connections that
is considered \(lqslow\(rq for the purpose of the limit modifier options.

.SS "Limit modifiers"
The following arguents may be specified after the \fBExecution limit\fR
arguments to modify the value of the limit under specified circumstances,
for example the machine running on battery, running without a network connection,
etc.

.TP
\fB-batt\fR \fImod\fR, \fB--on-battery\fR \fImod\fR
Change the value of the preceding \fBExecution limit\fR by the value
of \fImod\fR in case that the machine is running on battery.

.TP
\fB-slnw\fR \fImod\fR, \fB--slow-nw\fR \fImod\fR
Change the value of the preceding \fBExecution limit\fR by the value
of \fImod\fR in case that the current combined rated speed of all network
interfaces is less than the limit specified by the \fB-snw\fR option.

.TP
\fB-nonw\fR \fImod\fR, \fB--no-nw\fR \fImod\fR
Change the value of the preceding \fBExecution limit\fR by the value
of \fImod\fR in case that there is no network connection.

.SH "EXAMPLES"
.EX
.B atmost -n 8 -- cmd args...
.EE
.PP
If invoked multiple times concurrently, allows at most 8 instances
of \(lqcmd\(rq to run at the same time. If there are more invocations
they wait until some of the already running instances finish.

.EX
.B atmost -l 10 -batt -4 -- cmd args...
.EE
.PP
If not running on battery and the 1 minute average CPU load is more than
10% or if running on battery and the 1 minute average CPU load is more
than 6% (10 + \-4) then wait for until the load drops under the specified
limits and run the command after \fB--\fR, otherwise run the command
immediately.

.EX
.B atmost -snw 10 -m 30 -io 100 -batt -40 -slnw -20  -- cmd args...
.EE
.PP
Waits if there is less than 30% available memory or if not running on battery
if there is less then 100 current I/O operations. If running on battery then
lower the limit of concurrent I/O operations by 40. If there is a network
connection slower then 10 Mb/s, then lower the limit of I/O operations
by 20. The limit modifiers do stack.

.SH "SEE ALSO"
.BR atmost_driver (1),

.SH "AUTHOR"
Matus Chochlik, chochlik@gmail.com
.SH "COPYRIGHT"
Copyright (c) 2019 Matus Chochlik
.PP
Permission is granted to copy, distribute and/or modify this document
under the terms of the Boost Software License, Version 1.0.
(See a copy at http://www.boost.org/LICENSE_1_0.txt)


python/atmost.py
#!/usr/bin/env python
#coding=utf8

from __future__ import print_function

import os
import sys
import argparse
import subprocess
import posix_ipc
import hashlib

# ------------------------------------------------------------------------------
def print_error(error):
    print("atmost error: %s" % (str(error)), file=sys.stderr)
#------------------------------------------------------------------------------#
def which(executable):
    path=os.getenv('PATH')
    for path in path.split(os.path.pathsep):
        path = os.path.join(path,executable)
        if os.path.exists(path) and os.access(path, os.X_OK):
            return os.path.realpath(path)
#------------------------------------------------------------------------------#
class __AtMostArgumentParser(argparse.ArgumentParser):

    # --------------------------------------------------------------------------
    def __init__(self, **kw):
        argparse.ArgumentParser.__init__(self, **kw)

        def _positive_int(s):
            try:
                n = int(s)
                assert(n > 0)
                return n
            except:
                self.error("`%s' is not a positive integer value" % str(x))

        self.add_argument(
            '-r', '--reset',
            dest='reset',
            default=False,
            action='store_true'
        )

        self.add_argument(
            '-n', '--limit',
            dest='limit',
            metavar='PROCESS-COUNT',
            nargs='?',
            type=_positive_int,
            default=None
        )

        self.add_argument(
            dest="command_line",
            nargs=argparse.REMAINDER,
            default=list(),
            help="""The command to execute."""
        )

    # --------------------------------------------------------------------------
    def process_parsed_options(self, options):

        if options.limit is None:
            options.limit = 2
            try:
                import multiprocessing
                options.limit = multiprocessing.cpu_count()
            except ModuleNotFoundError: pass
            except ImportError: pass

        return options


    # --------------------------------------------------------------------------
    def parse_args(self):
        return self.process_parsed_options(
            argparse.ArgumentParser.parse_args(self)
        )

# ------------------------------------------------------------------------------
def make_argparser():
    return __AtMostArgumentParser(
        prog="atmost",
        description="""
        Limits the number of cuncurrently running instances of a specific
        executable
        """
    )
# ------------------------------------------------------------------------------
def execute(options):
    try:
        exe_path = which(options.command_line[0])
        if exe_path is None:
            print_error("command '%s' not found" % (options.command_line[0]))
            return 2
    except IndexError as index_error:
        print_error("empty command line given")
        return 1

    sem_name = '/' + hashlib.md5(exe_path.encode()).hexdigest()

    if options.reset:
        posix_ipc.unlink_semaphore(sem_name)

    try:
        semaphore = posix_ipc.Semaphore(
            name = sem_name,
            flags = posix_ipc.O_CREAT,
            initial_value = options.limit
        )
        semaphore.acquire();

        result = 100
        try:
            result = subprocess.call(options.command_line)
        except Exception as error:
            print_error(error)
    finally:
        semaphore.release();
        semaphore.close()

    return result
# ------------------------------------------------------------------------------
def main():
    return execute(make_argparser().parse_args())
# ------------------------------------------------------------------------------
if __name__ == "__main__":
    exit(main())
